pkg/cache/stats_counter.go

```diff
@@ -0,0 +1,229 @@
+			Name: "cache_hits_total",
+			Help: "Total number of cache hits.",
+		}),
+		miss: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cache_miss_total",
+			Help: "Total number of cache misses.",
+		}),
+		eviction: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cache_evictions_total",
+			Help: "Total number of cache evictions.",
+		}),
+	}
+	for _, opt := range options {
+		opt(s)
+	}
+	return s
+}
+
+// RecordHits records the number of hits.
+// This method is called by Get and GetIfPresent methods on a cache hit.
+func (c *burrowStatsCounter) RecordHits(hits uint64) {
+	c.hits.Add(float64(hits))
+}
+
+// RecordMisses records the number of misses.
+// This method is called by Get and GetIfPresent methods method on a cache miss.
+func (c *burrowStatsCounter) RecordMisses(miss uint64) {
+	c.miss.Add(float64(miss))
+}
+
+// RecordLoadSuccess records the number of successful loads.
+// This method is called by Get methods on a successful load.
+func (c *burrowStatsCounter) RecordLoadSuccess(loadTime time.Duration) {
+	if !c.trackLoadingCacheStats {
+		return
+	}
+	c.loadSuccess.Inc()
+	c.loadTotalTime.Observe(loadTime.Seconds())
+}
+
+// RecordLoadError records the number of failed loads.
+// This method is called by Get methods on a failed load.
+func (c *burrowStatsCounter) RecordLoadError(loadTime time.Duration) {
+	if !c.trackLoadingCacheStats {
+		return
+	}
+	c.loadError.Inc()
+	c.loadTotalTime.Observe(loadTime.Seconds())
+}
+
+// RecordEviction records the number of evictions.
+func (c *burrowStatsCounter) RecordEviction() {
+	c.eviction.Inc()
+}
+
+// Snapshot records the current stats.
+// This method is called only by Stats method.
+// And Stats method is just for debugging purpose.
+func (c *burrowStatsCounter) Snapshot(s *burrow.Stats) {
+	var err error
+	s.HitCount, err = getCurrentCounterValue(c.hits)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to collect cache hits", "err", err)
+	}
+	s.MissCount, err = getCurrentCounterValue(c.miss)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to collect cache misses", "err", err)
+	}
+	s.EvictionCount, err = getCurrentCounterValue(c.eviction)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to collect cache evictions", "err", err)
+	}
+
+	if !c.trackLoadingCacheStats {
+		return
+	}
+	s.LoadSuccessCount, err = getCurrentCounterValue(c.loadSuccess)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to collect cache load success", "err", err)
+	}
+	s.LoadErrorCount, err = getCurrentCounterValue(c.loadError)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to collect cache load error", "err", err)
+	}
+	totalTime, err := getCurrentHistogramSumValue(c.loadTotalTime)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to collect cache load time", "err", err)
+	}
+	s.TotalLoadTime = time.Duration(totalTime)
+}
+
+func collect(col prometheus.Collector) (*dto.Metric, error) {
```

Comment:
this `collect` method still has the same issue as the other functions had before I think